src/com/carolinarollergirls/scoreboard/json/ScoreBoardJSONListener.java

```diff
@@ -280,6 +285,26 @@ private void processPosition(String path, Position p, boolean remove) {
 		updateMap.put(path + "." + Position.EVENT_SKATER, p.getSkater() == null ? null : p.getSkater().getId());
 		updateMap.put(path + "." + Position.EVENT_PENALTY_BOX, p.getPenaltyBox());
 	}
+	
+	private void processPenaltyCodes(Settings s) {
+		PenaltyCodesManager pm = new PenaltyCodesManager(s);
+		updateMap.put("ScoreBoard.PenaltyCode", null);
```

Comment:
I don't think you can both clear a subtree like this and give it entries in one go, as the updateMap can be processes in any order as maps are unordered.

Reply:
Since updateMap is a LinkedHashMap, the iteration order is stable in insertion order.

Reply:
There's two problems with that. Firstly a entry could be nulled and later have a value put in, and the subtree wouldn't get cleared. Secondly we don't iterate over the updates in WS.updateState, we get the set of keys and iterate over that.

We'd need to change that code somehow to make this work correctly in general.

Reply:
> Firstly a entry could be nulled and later have a value put in, and the subtree wouldn't get cleared.
I agree on this in the general case, I think the subtree deletion 

> Secondly we don't iterate over the updates in WS.updateState, we get the set of keys and iterate over that.

Reply:
> Firstly a entry could be nulled and later have a value put in, and the subtree wouldn't get cleared.

I agree on this in the general case, I think the subtree state deletion basically only works on accident.  In this case though I'm never putting anything into the top-level key.  I'd probably rather update WS to provide an explicit "removeState/deleteState" function to give us more explicit control over what is in the state object.  

If you have an alternative approach in mind, we can break out that discussion to the FB or slack.  Unless you think it's too big of a change for this PR, I can incorporate it with these changes.

> Secondly we don't iterate over the updates in WS.updateState, we get the set of keys and iterate over that.

It's original-key-insertion-order specifically.  That's definitely more nuanced than just insertion order, but since the updateMap is emptied after handing off to WS.updateState, that doesn't become a problem in practice all that much.

Reply:
Pondering this more, this design as-is is just not safe in general. A hash map is the wrong data structure, as it loses ordering. This may not be an issue for a single deletion and then adding it back, however there's the notion of batching so you could have delete+add+delete+add which would end up with a mixture of the two adds rather than only the last add being present.

I don't believe existing code is affected, as skaters don't change attributes from update to update; and the set of policies is fixed. By contrast penalty codes can be added and removed.

Without getting into a full-transactional design, I think switching the updateMap to a list should do the trick. That way the updates always happen in order, and will still get squashed down before being sent out to clients.